legacy/builder/phases/core_builder.go

```diff
@@ -121,6 +122,7 @@ func compileCore(ctx *types.Context, buildPath *paths.Path, buildCachePath *path
 	// archive core.a
 	if targetArchivedCore != nil && !ctx.OnlyUpdateCompilationDatabase {
 		err := archiveFile.CopyTo(targetArchivedCore)
+		buildcache.GetOrCreate(targetArchivedCore.Parent().Parent(), targetArchivedCore.Base())
```

Comment:
This call needs to be moved up to line 100. In the current location, the call is never reached after the core has been cached because the function returns early in that case:
https://github.com/arduino/arduino-cli/blob/c12bb0d7154dc32536384bdba32458787229bcb7/legacy/builder/phases/core_builder.go#L108

Reply:
I moved it there. It also creates the directory if missing, so maybe we can avoid the double call to mkdirall?

Reply:
My only concern is there is no error handling on the `buildcache.GetOrCreate` call, so if for some reason that call doesn't create the directory then the user would get an inappropriate error message when Arduino CLI attempts to cache the core:

https://github.com/arduino/arduino-cli/blob/3989980c11976f702bba059327dabc0d2f71ab66/legacy/builder/phases/core_builder.go#L130-L132

Reply:
In which scenario does the user gets a `NotExist` now?

Reply:
They will get it if the command generated from the `recipe.ar.pattern` property defined in the `platform.txt` file of the platform of the board they are compiling does not generate a core archive file, or does not place it at the location Arduino CLI's caching system expects to find it at:

https://github.com/arduino/arduino-cli/blob/165789cf/legacy/builder/phases/core_builder.go#L118

The boards platform developer can define any arbitrary command patterns, so it is possible for the compilation to work perfectly well but not integrate correctly into Arduino CLI's caching system. Here is a silly example of a modification to [the `arduino:avr` platform's](https://github.com/arduino/arduinocore-avr) configuration that would still allow compilation to work fine, but break core caching:

```patch
diff --git a/platform.txt b/platform.txt
index df5ddb7..137a249 100644
--- a/platform.txt
+++ b/platform.txt
@@ -63,10 +63,10 @@ recipe.S.o.pattern="{compiler.path}{compiler.c.cmd}" {compiler.S.flags} -mmcu={b
 ## Create archives
 # archive_file_path is needed for backwards compatibility with IDE 1.6.5 or older, IDE 1.6.6 or newer overrides this value
 archive_file_path={build.path}/{archive_file}
-recipe.ar.pattern="{compiler.path}{compiler.ar.cmd}" {compiler.ar.flags} {compiler.ar.extra_flags} "{archive_file_path}" "{object_file}"
+recipe.ar.pattern="{compiler.path}{compiler.ar.cmd}" {compiler.ar.flags} {compiler.ar.extra_flags} "{build.path}/foo.a" "{object_file}"
 
 ## Combine gc-sections, archives, and objects
-recipe.c.combine.pattern="{compiler.path}{compiler.c.elf.cmd}" {compiler.c.elf.flags} -mmcu={build.mcu} {compiler.c.elf.extra_flags} {compiler.ldflags} -o "{build.path}/{build.project_name}.elf" {object_files} {compiler.libraries.ldflags} "{build.path}/{archive_file}" "-L{build.path}" -lm
+recipe.c.combine.pattern="{compiler.path}{compiler.c.elf.cmd}" {compiler.c.elf.flags} -mmcu={build.mcu} {compiler.c.elf.extra_flags} {compiler.ldflags} -o "{build.path}/{build.project_name}.elf" {object_files} {compiler.libraries.ldflags} "{build.path}/foo.a" "-L{build.path}" -lm
 
 ## Create output files (.eep and .hex)
 recipe.objcopy.eep.pattern="{compiler.path}{compiler.objcopy.cmd}" {compiler.objcopy.eep.flags} {compiler.objcopy.eep.extra_flags} "{build.path}/{build.project_name}.elf" "{build.path}/{build.project_name}.eep"
```

I don't know the exact details, but I believe the problem was that there were already a lot of 3rd party boards platforms in the wild before the caching system was implemented. At the time those platforms were created, there was no reason for the platform developer to put a core archive in a specific location. So the system was designed to continue with the compilation, which is not dependent on the ability to create a cache, but display this warning to the users so they would report to the platform developers that their platform configuration did not allow the beneficial caching to occur.

---

Information about `recipe.ar.pattern` here:

https://arduino.github.io/arduino-cli/dev/platform-specification/#recipes-to-build-the-corea-archive-file